### ISSUE TITLE
Improve foreign key property/referenced relation coexistence

### ISSUE THREAD

**Issue by jaredjj3** [mikro-orm/mikro-orm#2099]
**Is your feature request related to a problem? Please describe.**

TL;DR - I would like to establish relationships using identified references or direct foreign key assignments.

#1469 is an issue where a user had a use case to map 2 entity properties to 1 column. A partial quote from [this comment](https://github.com/mikro-orm/mikro-orm/issues/1469#issuecomment-783188128):

>Nope, this is not supported, you can't map two properties to single column - only if you make one of them virtual (so it will be ignored during persistence as well as schema generation).
>
>I never understood why people want this.

However, there can be database-independent contexts (e.g. a separate cache or static assignment) that can store a foreign key. I would like to be able to establish a relationship using either an identified reference or a foreign key. I think this is a legitimate use case and a reason why people would want this.

The solution in #1469 only partially works. You can _get_ a foreign key via the reference, but you cannot _set_ it.

I created a MikroORM an easy-to-run test playground repo in [app-example](https://github.com/jaredjj3/app-example). The latest commit at the time of this FR is [23eec78cafb7c016222e19a928ed484ee25e604b](https://github.com/jaredjj3/app-example/tree/23eec78cafb7c016222e19a928ed484ee25e604b).

The solution I came up with the existing library is [here](https://github.com/jaredjj3/app-example/blob/23eec78cafb7c016222e19a928ed484ee25e604b/src/entities/Post.ts#L28-L45). The test cases I'm covering is [here](https://github.com/jaredjj3/app-example/blob/23eec78cafb7c016222e19a928ed484ee25e604b/src/MikroORM.test.ts#L87). The problem with this solution is that it hacks the `__initialized` property and calling `Reference.load` (rightfully) throws an error:

```
[ValidationError: Entity User is not managed. An entity is managed if its fetched from the database or registered as new through EntityManager.persist()]
```

Callers must clear the reference in the cache's identity map in order to load as expected. This aspect is not blocking me, but remembering to do this is undesired.

**Describe the solution you'd like**

A `@ForeignKey` decorator (or `@Property` configuration option) that links a foreign key property with a relationship reference. This would allow establishing a relationship using either an identified reference or a foreign key. Under-the-hood, it could create a getter and setter similar to what I came up with.

```ts
@Entity()
class Post {
  @ManyToOne(() => User, { wrappedReference: true })
  @IsNotEmpty()
  author?: IdentifiedReference<User, 'id'>;
  
  @ForeignKey(() => Post, (post) => post.author)
  authorId?: number;
}
```

```ts
const author = await em.findOne(User, {});

const post1 = new Post();
post1.author = wrap(author).toReference();  // this works and is already supported

const post2 = new Post();
post2.authorId = author.id;  // I would like this to work without hacking __initialized
```

**Describe alternatives you've considered**

I considered using `em.nativeCreate` and `em.nativeUpdate`, but it's error-prone to remember what lifecycle hooks to also call.

I also considered accepting that you cannot assign directly by foreign key, and that referenced entities should be loaded before assignment. However, this added a lot of complexity in some contexts, so I decided a FR was worth a try.

**Additional context**

To run the tests in the example repo:

1. Clone the [example repo](https://github.com/jaredjj3/app-example/).
1. cd to the project directory
1. Read the README and ensure the prerequisites are met.
1. Run `yarn start` to run the tests.

In the example repo, all entities inherit from the [Base class](https://github.com/jaredjj3/app-example/blob/23eec78cafb7c016222e19a928ed484ee25e604b/src/entities/Base.ts). This class declares a `BeforeCreate` and `BeforeUpdate` hook that ensures the entity is valid at the application level before attempting to persist it. The actual validation is performed by the [class-validator](https://github.com/typestack/class-validator) library.

**Comment by jaredjj3** [comment]
In the example repo, I isolated the hack in this [module](https://github.com/jaredjj3/app-example/blob/36a2a9b8819bf4e19537d6dc0885ba75485d27be/src/hacks/HACK_ISSUE_2099.ts) and I am somewhat successfully using it in the [foreign key setter](https://github.com/jaredjj3/app-example/blob/36a2a9b8819bf4e19537d6dc0885ba75485d27be/src/entities/Post.ts#L37).

I added a [failing test case](https://github.com/jaredjj3/app-example/blob/36a2a9b8819bf4e19537d6dc0885ba75485d27be/src/MikroORM.test.ts#L147-L168) for the desired behavior. The test fails in this manner: 

![image](https://user-images.githubusercontent.com/19232300/128646745-c0a3deaf-e955-4023-bf8e-0f1e7c7019bc.png)

I found in a [different test case](https://github.com/jaredjj3/app-example/blob/36a2a9b8819bf4e19537d6dc0885ba75485d27be/src/MikroORM.test.ts#L170-L195) that I could get the behavior I want by leveraging `{ refresh: true }` in the next `em.find*` call that includes the entity in question. If I could mark an entity/reference as stale, I think this would solve the problem without being too invasive.

**Comment by B4nan** [comment]
I don't plan to support this, for reasons mentioned in #1469. In fact I don't think this is possible with current implementation, not consistently as for not managed entities we do not know which EM to use for the reference - references need to be part of the identity map. 

What I would like to try is to make the `refresh` option more loose, as it might not be needed if we can verify the entity was not changed. But it will require computing change set on each, which might have some performance impact, but it would be probably worth the improved DX. But it might as well be too complex or just add more confusion, as it would basically disable the identity map invariant - right now it needs to be explicit which always adds to readability. 

> I also considered accepting that you cannot assign directly by foreign key, and that referenced entities should be loaded before assignment. However, this added a lot of complexity in some contexts, so I decided a FR was worth a try.

You can assign by a FK, you just need the entity to be managed, and use `wrap().assign()` to perform it. It will handle mapping to the reference object.

```ts
const author = await em.findOne(User, 1);

// post1 is not managed, we need to provide the `em`
const post1 = new Post();
wrap(post1).assign({ author: author.id }, { em });

// post2 is managed, no need for explicit `em` option
const post2 = await em.findOne(Post, 2);
wrap(post2).assign({ author: author.id });
```

> In the example repo, I isolated the hack in this module and I am somewhat successfully using it in the foreign key setter.

Your hack won't work properly in all cases, to set a reference, you need to know what EM is managing it. For managed entity we could get it via `wrap(this, true).__em`, but for not managed entities this is impossible to achieve. I would expect that it will be also impossible to load such relation, as it won't be managed by the right entity manager (actually it looks won't be managed at all). 

> I added a failing test case for the desired behavior. The test fails in this manner:

Due to how identity map works, both `post` and `actualPost` as the exact same objects (`post === actualPost`). You need `refresh: true` in that exact case if you want to reload the entity - that is the flag that will make it so they are not equal - that the entity in the identity map will be refreshed with the new one. Alternative is to use `em.populate()` instead.

**Comment by jaredjj3** [comment]
Thank you for the comprehensive response. I understand that the complexity overhead is not worth it.

This solution works for me:

```ts
wrap(post).assign({ author: user.id }, { em });
```

I didn't realize you can assign references like this. This is covered in the [docs](https://mikro-orm.io/docs/entity-helper/#updating-entity-values-with-entityassign), but I failed to understand it.

Now that I understand it, I made a solution that doesn't hack the internals of entities:

```ts
export type BaseOpts = {
  em?: EntityManager;
};

export abstract class Base {
  em?: EntityManager;

  constructor(opts: BaseOpts = {}) {
    this.em = opts.em;
  }
  
  // ...

}
```

```ts
@Entity({ tableName: 'posts' })
export class Post extends Base {

  // ...
  
  @ManyToOne(() => User, { wrappedReference: true })
  @IsNotEmpty()
  author?: IdentifiedReference<User, 'id'>;
  
  @Property({ persist: false })
  get authorId() {
    return this.author?.id;
  }

  set authorId(authorId: number) {
    if (this.authorId === authorId) {
      return;
    }
    if (!this.em) {
      throw new Error('must assign EntityManager to assign by foreign key');
    }
    wrap(this).assign({ author: authorId }, { em: this.em });
  }
  
  constructor(props: Partial<Post> = {}, opts: BaseOpts = {}) {
    super(opts);
    Object.assign(this, props);
  }
}
```

The only downside I can think of at the moment is unknown behavior using multiple entity managers in a given context, but I don't think that will be an issue for my use case. I decided not to mess with `wrap(this, true).__em` based on the guidance from the [docs](https://mikro-orm.io/docs/lifecycle-hooks/#limitations-of-lifecycle-hooks):

>The internal instance of EntityManager accessible under wrap(this, true).__em is not meant for public usage.

I'm going to try to figure out if I can make a decorator that creates `get` and `set` property descriptors like this, and post it here for posterity and for other people that want to achieve this.

**Comment by jaredjj3** [comment]
Here's is a rough `ForeignKey` decorator. It assumes that the foreign key points to the 'id' column of the foreign table. I had some difficulty getting the types right, so I just casted the problematic parts to `any` and left the primary key column name not configurable:

_ForeignKey.ts_

```ts
import { wrap } from '@mikro-orm/core';
import { Base } from '../entities';

type Key = number | string;

export const ForeignKey = <T extends Base, K extends Key>(association: keyof T) => {
  return ((target: T, propertyKey: PropertyKey, descriptor: PropertyDescriptor = {}) => {
    const getKey = function (this: T): K {
      return (this[association] as any)?.id;
    };

    const setKey = function (this: T, key: K) {
      if (getKey.call(this) === key) {
        return;
      }
      if (!this.em) {
        throw new Error('must set EntityManager before assigning by foreign key');
      }
      wrap(this).assign({ [association]: key }, { em: this.em });
    };

    descriptor.get = getKey;
    descriptor.set = setKey;
    return descriptor;
  }) as any;
};

```

_Base.ts_

```ts
export type BaseOpts = {
  em?: EntityManager;
};

export abstract class Base {
  em?: EntityManager;

  constructor(opts: BaseOpts = {}) {
    this.em = opts.em;
  }
}
```

_Post.ts_

```ts
@Entity({ tableName: 'posts' })
export class Post extends Base {
  @ManyToOne(() => User, { wrappedReference: true })
  @IsNotEmpty()
  author?: IdentifiedReference<User, 'id'>;

  @Property({ persist: false })
  @ForeignKey<Post, number>('author')
  authorId!: number;
}
```

---

Example usage:

```ts
it('allows association using the foreign key', async () => {
  // Create a new user and clear cache to simulate a "clean slate".
  const user = new User({ username: rand.str(8) });
  em.persist(user);
  await em.flush();
  em.clear();
 
  const post = new Post({ title: rand.str(8), authorId: user.id }, { em });
  em.persist(post);
  await em.flush();

  await expect(em.count(Post)).resolves.toBe(1);
  await expect(em.count(User)).resolves.toBe(1);
});

it('disallows association using the foreign key without entity manager', async () => {
  const user = new User({ username: rand.str(8) });
  em.persist(user);
  await em.flush();
  em.clear();

  expect(() => new Post({ title: rand.str(8), authorId: user.id })).toThrowError();
});
```
